### PR TITLE
Add a rule to prevent facade framework assemblies

### LINKED ISSUE #30: Add rule to check for bad net451 references
Add a rule to check for bad references to prevent issues like these:

* https://github.com/aspnet/EntityFramework/issues/4708
* https://github.com/aspnet/Logging/pull/376

### PR DESCRIPTION
**Pull request by bricelam**
Resolves #30

### CODE REVIEW

**Comment by bricelam** [comment]
@ajaybhargavb which list do I add this to? The one in `Program`, `AdxVerificationCompositeRule`,  `DefaultCompositeRule`, and/or somewhere else?

File: src/NuGetPackageVerifier/PackageIssueFactory.cs [hunk 1]
```diff
@@ -184,5 +187,19 @@ public static PackageVerifierIssue AssemblyHasWrongJsonNetVersion(string assembl
...
                 MyPackageIssueLevel.Error);
         }
+
+        public static PackageVerifierIssue FrameworkAssembliesContainFacade(string assemblyName, FrameworkName supportedFramework)
+        {
+            var shortFrameworkName = supportedFramework.GetShortName();
+
+            return new PackageVerifierIssue(
+                "NUSPEC_FACADE",
+                shortFrameworkName,
```
Comment (bricelam) [inline src/NuGetPackageVerifier/PackageIssueFactory.cs]: Looks like `instance` needs to be a little more distinct. :smile:

File: src/NuGetPackageVerifier/Extensions/FrameworkNameExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+            if (frameworkNameProvider.TryGetShortIdentifier(frameworkName.Identifier, out identifierShortName)
+                && frameworkNameProvider.TryGetShortProfile(
+                    frameworkName.Identifier,
+                    frameworkName.Profile,
+                    out profileShortName))
+            {
+                return FrameworkNameHelpers.GetFolderName(
+                    identifierShortName,
+                    frameworkNameProvider.GetVersionString(frameworkName.Version),
+                    profileShortName);
```
Comment (bricelam) [inline src/NuGetPackageVerifier/Extensions/FrameworkNameExtensions.cs]: Eww, this results in `net451-Full`. Will fix.

**Comment by ajaybhargavb** [comment]
@bricelam, yes you need to add it in all three that you mentioned.

**Comment by ajaybhargavb** [comment]
This code needs to change considerably because I am currently working on updating this project to use `NuGet,Packaging` instead of `NuGet.Core`

**Comment by bricelam** [comment]
@ajaybhargavb I'm happy to rebase this PR once that's merged. I've ran this locally to find the packages I need to fix, so it doesn't need to be merged right away.

**Comment by ajaybhargavb** [comment]
Great, thanks :smile:

File: src/NuGetPackageVerifier/Rules/FrameworkAssembliesDoesNotContainFacadesRule.cs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+            "System.Xml.XDocument",
+            "System.Xml.XmlSerializer"
+        };
+
+        public IEnumerable<PackageVerifierIssue> Validate(
+            IPackageRepository packageRepo,
+            IPackage package,
+            IPackageVerifierLogger logger)
+        {
+            return from a in package.FrameworkAssemblies
```
Comment (Eilon) [inline src/NuGetPackageVerifier/Rules/FrameworkAssembliesDoesNotContainFacadesRule.cs]: Wow, someone used actual LINQ syntax. Bravo :smile:

Comment (bricelam) [inline src/NuGetPackageVerifier/Rules/FrameworkAssembliesDoesNotContainFacadesRule.cs]: Lol, I try to use comprehension syntax everywhere it's more terse. I've even developed strange patterns like this to keep it readable:
```C#
var list = Enumerable.ToList(
    from t in things
    select t);
```

Comment (Eilon) [inline src/NuGetPackageVerifier/Rules/FrameworkAssembliesDoesNotContainFacadesRule.cs]: Ah, I used to do `(<linq stuff>).ToList()` but I like that more. But I like just the plain old method calls the most :smile: To each their own.

**Comment by ajaybhargavb** [comment]
@bricelam, https://github.com/aspnet/DnxTools/pull/32 is in.

**Comment by bricelam** [comment]
Rebased.

**Comment by ajaybhargavb** [comment]
:shipit: and make sure to react to any repo that fails the check.

**Comment by bricelam** [comment]
I'm trying to be proactive. :wink: I'll merge all my fixes before this.

**Comment by ajaybhargavb** [comment]
Even better :smile:

File: src/NuGetPackageVerifier/Rules/FrameworkAssembliesDoesNotContainFacadesRule.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+using System.IO;
+using System.Linq;
+using NuGet.Frameworks;
+using NuGet.Packaging;
+using NuGetPackageVerifier.Logging;
+
+namespace NuGetPackageVerifier.Rules
+{
+    public class FrameworkAssembliesDoesNotContainFacadesRule : IPackageVerifierRule
+    {
```
Comment (davidfowl) [inline src/NuGetPackageVerifier/Rules/FrameworkAssembliesDoesNotContainFacadesRule.cs]: You could do it better by actually looking at the facades folder

Comment (bricelam) [inline src/NuGetPackageVerifier/Rules/FrameworkAssembliesDoesNotContainFacadesRule.cs]: :smile: @Eilon was concerned about speed and flakiness, so I kept it super simple.

**Pull request closed without merging**